Add MwaaTaskSensor to Amazon Provider Package

seanghaeli opened the pull request:
The `MwaaTaskSensor` waits for the completion of a DAG task instance in an MWAA environment. This PR includes an implementation with unit tests, system tests, and docs. Similar to [MwaaDagRunSensor](https://github.com/apache/airflow/pull/46945)

Also modified system test to have `MwaaTriggerDagRunOperator` set to `deferrable=True`. This tests the `MwaaTaskSensor` and `MwaaDagRunSensor` sensors _during_ execution of DAG Run rather than only afterwards.

seanghaeli committed:
Add MwaaTaskSensor to Amazon Provider Package

seanghaeli committed:
include pre-commit hooks

eladkal commented on providers/amazon/src/airflow/providers/amazon/aws/triggers/mwaa.py:
```diff
@@ -106,6 +106,87 @@ def hook(self) -> AwsGenericHook:
+        failure_states: Collection[str] | None = None,
+        waiter_delay: int = 60,
+        waiter_max_attempts: int = 720,
+        aws_conn_id: str | None = None,
```
This isn't consistent with what was discussed on https://github.com/apache/airflow/pull/51196 is it?

seanghaeli replied:
Updated the logic here for review

eladkal commented on providers/amazon/src/airflow/providers/amazon/aws/triggers/mwaa.py:
```diff
@@ -106,6 +106,87 @@ def hook(self) -> AwsGenericHook:
+        if len(self.success_states & self.failure_states):
+            raise ValueError("success_states and failure_states must not have any values in common")
+
+        in_progress_states = {s.value for s in TaskInstanceState} - self.success_states - self.failure_states
```
is this logic right? If we fall back to the defaults it will consider `skipped` and `removed` as in progress state.
Also, there is no protection here against possible future addition of new state to task instance. For example we are discussing https://github.com/apache/airflow/issues/12199

I suggest to add defensive test around adding more states so we'll know to modify code here or maybe we can consider adding more classes to categorized states similar to https://github.com/apache/airflow/blob/083e03a909f923527d9d2f8d978962ddfb6e5b7a/task-sdk/src/airflow/sdk/api/datamodels/_generated.py#L415-L419

o-nikolas replied:
There is also `upstream_failed` for example and others that are terminal that will cause this to wait forever right?

seanghaeli replied:
Updated the logic here for review

eladkal commented on providers/amazon/src/airflow/providers/amazon/aws/sensors/mwaa.py:
```diff
@@ -159,3 +159,137 @@ def execute(self, context: Context):
+                    external_task_id=self.external_task_id,
+                    success_states=self.success_states,
+                    failure_states=self.failure_states,
+                    # somehow the type of poke_interval is derived as float ??
```
please clarify? We don't normally leave such todos in the code.

eladkal commented on providers/amazon/src/airflow/providers/amazon/aws/sensors/mwaa.py:
```diff
@@ -159,3 +159,137 @@ def execute(self, context: Context):
+                    success_states=self.success_states,
+                    failure_states=self.failure_states,
+                    # somehow the type of poke_interval is derived as float ??
+                    waiter_delay=self.poke_interval,  # type: ignore[arg-type]
```
why do we need the type ignore? this may hide an issue. I don't see we have such ignores in any of the other cases in the code.
Maybe it just requires
`waiter_delay=int(self.poke_interval)`
?

eladkal commented on providers/amazon/tests/system/amazon/aws/example_mwaa.py:
```diff
@@ -66,6 +66,22 @@ def unpause_dag(env_name: str, dag_id: str):
+        env_name=env_name, path=f"/dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances", method="GET"
+    )
+    print("test debug")
+    print(response)
```
leftovers?

eladkal commented on providers/amazon/docs/operators/mwaa.rst:
```diff
@@ -76,6 +76,24 @@ In the following example, the task ``wait_for_dag_run`` waits for the DAG run cr
+Wait on the state of an AWS MWAA Task
+========================================
+
+To wait for a Task Instance within a DAG Run running on Amazon MWAA until it reaches one of the given states, you can use the
```
We should clarify that this is intended to be used for sensing a task executed on a different MWAA environment.

also, the title is `Task` yet the description is `DAG Run`. It's not the same.
Does the sensor offer both capabilities?

ramitkataria commented on providers/amazon/src/airflow/providers/amazon/aws/triggers/mwaa.py:
```diff
@@ -106,6 +106,87 @@ def hook(self) -> AwsGenericHook:
+        *,
+        external_env_name: str,
+        external_dag_id: str,
+        external_dag_run_id: str,
```
I think we should also make `external_dag_run_id` optional so that users don't have to manually specify the dag run if they just want to use the latest run

seanghaeli replied:
What would be the use case for this? Is there a situation where the user knows the Task Instance ID but not the Dag Run ID? For example, in the test case right now, I only am able to fetch the Task Instance ID by first knowing the Dag Run ID.

ramitkataria replied:
[Discussed offline] task ID is set while authoring a Dag (so the user should know the task ID of the task they want to wait for) but Dag ID is generated when the user runs the Dag

ramitkataria commented on providers/amazon/tests/system/amazon/aws/example_mwaa.py:
```diff
@@ -66,6 +66,22 @@ def unpause_dag(env_name: str, dag_id: str):
 
+# Can only be run after 'trigger_dag_run' task is run.
+@task
+def get_task_id(env_name: str, dag_id: str):
```
What do you think about making the task ID an externally fetched variable similar to `EXISTING_DAG_ID_KEY`?

seanghaeli replied:
I'm not certain it would make sense. Take the current test case as an example:

1. The DAG name is passed in externally through `EXISTING_DAG_ID_KEY`.
2. `MwaaTriggerDagRunOperator` triggers a DAG Run.
3. An API call is made within the `get_task_id` function to find the task IDs within the above DAG Run.

If the DAG Run ID is not pased in externally, I'm not sure it makes sense for the Task instance ID to be. What do you think?

ramitkataria replied:
[Discussed offline and similar reasoning as above] task ID is set while authoring a Dag but Dag ID is generated when the user runs the Dag

seanghaeli committed:
- Comply with PR #51196: explicitly pass  to its superclass, and adjust the default value of  in base class to .
- Add defensive test around adding more task instance states to keep  of the MwaaTaskCompletedTrigger up to date.
- Fix issue where  of the MwaaTaskSensor derives to  instead of  type.
- Modify documentation to clearly indicate that the MwaaTaskSensor is meant to sense tasks across different MWAA environments.
- Make  an optional parameter, where it defaults to the latest dag run.
- Externally fetch the task ID variable.
- Test the sensor while a DAG Run is still in progress.

o-nikolas commented on providers/amazon/src/airflow/providers/amazon/aws/sensors/mwaa.py:
```diff
@@ -132,7 +132,7 @@ def poke(self, context: Context) -> bool:
         if state in self.failure_states:
             raise AirflowException(
-                f"The DAG run {self.external_dag_run_id} of DAG {self.external_dag_id} in MWAA environment {self.external_env_name} "
+                f"The DAG run {self.external_dag_run_id} of DAG {self.external_dag_id} in MWAA environment {self.external_env_name}"
```
Why did you delete the space here? The env name and the word `failed` no longer have a space between them now? Or does the env name have a space included at the end of it already?

o-nikolas commented on providers/amazon/src/airflow/providers/amazon/aws/sensors/mwaa.py:
```diff
@@ -150,8 +150,148 @@ def execute(self, context: Context):
+        For more information on how to use this sensor, take a look at the guide:
+        :ref:`howto/sensor:MwaaTaskSensor`
+
+    :param external_env_name: The external MWAA environment name that contains the DAG Run you want to wait for
```
```suggestion
    :param external_env_name: The external MWAA environment name that contains the Task Instance you want to wait for
```

o-nikolas replied:
Here and below as well. This Operator just waits for a single task, not the whole Dag Run. I'm assuming this is just copy/paste from the above operator.

o-nikolas commented on providers/amazon/src/airflow/providers/amazon/aws/sensors/mwaa.py:
```diff
@@ -150,8 +150,148 @@ def execute(self, context: Context):
+        if state in self.failure_states:
+            raise AirflowException(
+                f"The task {self.external_task_id} of DAG run {self.external_dag_run_id} of DAG {self.external_dag_id} in MWAA environment {self.external_env_name}"
+                f"failed with state: {state}"
```
Same as comment above, no space?

o-nikolas commented on providers/amazon/src/airflow/providers/amazon/aws/triggers/base.py:
```diff
@@ -80,7 +80,7 @@ def __init__(
         waiter_max_attempts: int,
         waiter_config_overrides: dict[str, Any] | None = None,
-        aws_conn_id: str | None,
+        aws_conn_id: str | None = "aws_default",
```
I think this is probably a good change, but this is the base trigger and will affect all AWS triggers. So I'm curious what caused you to modify this one?

seanghaeli replied:
Prompted by @eladkal 's comment:

> This isn't consistent with what was discussed on #51196 is it?

In which the custom connection ID is passed through kwargs through the superclass. The AwsBaseOperator initializes the parameter as: `"aws_conn_id: str | None = "aws_default"`, whereas the AwsBaseWaiterTrigger does it like: `aws_conn_id: str | None`.

Let me know your thoughts. If you agree with the change, maybe the better thing to do is make it a separate PR?

o-nikolas commented on providers/amazon/src/airflow/providers/amazon/aws/triggers/mwaa.py:
```diff
@@ -87,14 +87,95 @@ def __init__(
+    """
+    Trigger when an MWAA Task is complete.
+
+    :param external_env_name: The external MWAA environment name that contains the DAG Run you want to wait for
```
```suggestion
    :param external_env_name: The external MWAA environment name that contains the Task Instance you want to wait for
```

Same as the Operator class, these param descriptions need slight updates for task waiting not dag run waiting.

seanghaeli committed:
documentation update

seanghaeli committed:
Response to PR #51719 comments

- Brought UPSTREAM_FAILED to a Terminal Task Instance State instead of an Intermediate State.
- Added REMOVED to the list of successful terminal task instance states.
- Iterate programmatically through successful, failure, and in progress states instead of hard-coding.

seanghaeli committed:
update tests

seanghaeli committed:
Remove duplicate pointer to mwaatasksensor docs

seanghaeli committed:
Merge branch 'main' into ghaeli/mwaa-task-sensor

seanghaeli committed:
merge #53000

seanghaeli committed:
Fix integration tests

seanghaeli committed:
removed unnecessary defensive tests for trigger acceptor states after #53000 merge

seanghaeli committed:
removed unnecessary defensive tests for trigger acceptor states after #53000 merge

seanghaeli committed:
remove State file from PR

vincbeck commented on providers/amazon/src/airflow/providers/amazon/aws/sensors/mwaa.py:
```diff
@@ -150,8 +150,148 @@ def execute(self, context: Context):
+        return state in self.success_states
+
+    def execute_complete(self, context: Context, event: dict[str, Any] | None = None) -> None:
+        validate_execute_complete_event(event)
```
We do not do any action besides validating the event. Is it needed? Another alternative to look at is, if validating the event is not necessary, we can end the task in the triggerer without having to go back to the worker using `end_from_trigger` attribute. [Documentation](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/deferring.html#exiting-deferred-task-from-triggers)

seanghaeli replied:
I see this validation logic in several operators and sensors, for example here in S3:

https://github.com/apache/airflow/blob/e7e89a07ff24d7942e1940d10b4da5be0013336e/providers/amazon/src/airflow/providers/amazon/aws/sensors/s3.py#L380-L390

I'm not certain when `event` could be `None`, but it seems that this check is done across many places in the Amazon provider. In this light, perhaps it is worth keeping this?

vincbeck replied:
It is done across many places but usually `execute_complete` does more than that right? In the example you mentioned in S3, we are doing more than `validate_execute_complete_event`. `end_from_trigger` is kind of new and not very known so maybe we have not done it because we did not know. Let see what others think

seanghaeli replied:
Added the `end_from_trigger` property and removed the `execute_complete` method from the sensor. The tests are passing

vincbeck replied:
Cool :)

seanghaeli replied:
I think the docs you mentioned

> [Documentation](https://airflow.apache.org/docs/apache-airflow/stable/authoring-and-scheduling/deferring.html#exiting-deferred-task-from-triggers)

may be out of date. It indicates I can define `method_name=None,` but in the main branch `method_name` must be a string:

https://github.com/apache/airflow/blob/6b618efa91b4ef80c4821537b30f14a49c2badb6/task-sdk/src/airflow/sdk/bases/operator.py#L1577

In fact, I can't find any implementation of `end_from_trigger` in the Airflow codebase except for specifically within the `DateTimeTrigger`.

vincbeck replied:
Interesting, from the [examples](https://github.com/apache/airflow/blob/main/providers/standard/src/airflow/providers/standard/sensors/date_time.py#L154) in the codebase, indeed you need to pass a method name to `method_name` even though this is not called ... I think this is a bad design (or I am misunderstanding something about this feature). Please free to either keep using `end_from_trigger` but providing a method name or revert back to what you had before.

vincbeck commented on providers/amazon/tests/system/amazon/aws/example_mwaa.py:
```diff
@@ -118,25 +121,52 @@ def test_iam_fallback(role_to_assume_arn, mwaa_env_name):
         external_dag_id=trigger_dag_id,
         external_dag_run_id="{{ task_instance.xcom_pull(task_ids='trigger_dag_run')['RestApiResponse']['dag_run_id'] }}",
         poke_interval=5,
+        deferrable=True,
```
Why hardcoding `deferrable` to `True`? We usually do not do that

seanghaeli committed:
Remove hard coding deferrable property

seanghaeli committed:
Remove unnecessary execute_complete function in sensor and instead use end_from_trigger

vincbeck reviewed (approved):
LGTM, some mypy failures to take care of though

seanghaeli committed:
Correctly use end_from_trigger attribute

seanghaeli committed:
Correctly use end_from_trigger attibute for both dag run sensor and task sensor

seanghaeli committed:
Remove unnecessary import
